jmbitcoin/jmbitcoin/secp256k1_transaction.py

```diff
@@ -388,57 +413,122 @@ def address_to_script(addr):
 # Output script to address representation
 
 def is_p2pkh_script(script):
+    """ Given a script as bytes, returns True if the script
+    matches the pay-to-pubkey-hash pattern (using opcodes),
+    otherwise returns False.
+    """
     if script[:3] == b'\x76\xa9\x14' and script[-2:] == b'\x88\xac' and len(
             script) == 25:
         return True
     return False
 
 def is_segwit_native_script(script):
-    """Is scriptPubkey of form P2WPKH or P2WSH"""
-    if script[:2] in [b'\x00\x14', b'\x00\x20']:
+    """Is script, as bytes, of form P2WPKH or P2WSH;
+    see BIP141 for definitions of 2 current valid scripts.
+    """
+    if not isinstance(script, bytes):
+        script = binascii.unhexlify(script)
+
+    if (script[:2] == b'\x00\x14' and len(script) == 22) or (
+        script[:2] == b'\x00\x20' and len(script) == 34):
         return True
     return False
 
 def script_to_address(script, vbyte=0, witver=0):
+    """ Given a hex or bytes script, and optionally a version byte
+    (for P2SH) and/or a witness version (for native segwit witness
+    programs), convert to a valid address (either bech32 or Base58CE).
+    An important tacit assumption: anything which does not match
+    the parsing rules of native segwit, or pay-to-pubkey-hash, is
+    assumed to be p2sh.
+    Note also that this translates scriptPubKeys to addresses, not
+    redeemscripts to p2sh addresses; for that, use p2sh_scriptaddr.
+    """
     if not isinstance(script, bytes):
         script = binascii.unhexlify(script)
     if is_segwit_native_script(script):
-        #hrp interpreted from the vbyte entry, TODO this should be cleaner.
+        #hrp interpreted from the vbyte entry, TODO: better way?
         if vbyte in [0, 5]:
             hrp = 'bc'
+        elif vbyte == 100:
+            hrp = 'bcrt'
         else:
             hrp = 'tb'
         return bech32addr_encode(hrp=hrp, witver=witver,
-                                 witprog=[ord(x) for x in script[2:]])
+        witprog=struct.unpack('{}B'.format(len(script[2:])), script[2:]))
```

Comment:
Try changing this so that the first argument to `struct.unpack` is bytes.

Reply:
Yes, thanks; at first I tried `'b{}B'.format(...` but of course that wasn't viable, format requires a string; can I cast to bytes in a way that's 2/3 compatible? bytes()? I guess I should just figure it out myself but in the middle of something else, if you know off the top of your head, let me know, but don't put yourself out otherwise :)

Reply:
Maybe just encode('ascii') the entire first argument to struct.unpack. Using format strings with struct.unpack seems strange here, it's not something I recall ever seeing before, maybe [struct](https://docs.python.org/3/library/struct.html) has a more suitable function?